pylossless/dash/mne_visualizer.py

```diff
@@ -142,69 +179,135 @@ def inst(self, inst):
-
-    def add_annot_shapes(self, annotations):
-        self.layout.shapes = [self._get_annot_shape(annot)
-                              for annot in annotations]
-        self.layout.annotations = [self._get_annot_text(annot)
-                                   for annot in annotations]
-        if self.annotating:
-            self.layout.shapes += (self.annotation_inprogress,)
-
-    def refresh_annotations(self):
+    ###########################################################
+    # Methods for converting MNE annots to Plotly Shapes/annots
+    ###########################################################
+    def _get_annot_text(self, annotation, name=None):
+        """Get description from mne.annotation.
+
+        Parameters
+        ----------
+        annotation : mne.annotation
+            an mne.annotation from a raw object.
+        returns
+        -------
+        a dict that can be used as a plotly annotation (text).
+        """
+        return go.layout.Annotation(
+                    dict(x=annotation['onset'] + annotation['duration'] / 2,
+                         y=self.layout.yaxis['range'][1],
+                         text=annotation['description'],
+                         name=name,
+                         showarrow=False,
+                         yshift=10,
+                         font={'color': '#F1F1F1'}))
+
+    def _get_annot_shape(self, annotation, name='description'):
+        """Make a plotly shape from an mne.annotation."""
+        editable = True if 'bad' in annotation['description'] else False
+        opacity = 0.51 if annotation['duration'] else .75
+        layer = "below" if annotation['duration'] else 'above'
+        x1 = annotation['onset'] + annotation['duration']
+        return go.layout.Shape(dict(name=name,
+                                    editable=editable,
+                                    type="rect",
+                                    xref="x",
+                                    yref="y",
+                                    x0=annotation['onset'],
+                                    y0=self.layout.yaxis['range'][0],
+                                    x1=x1,
+                                    y1=self.layout.yaxis['range'][1],
+                                    fillcolor='red',
+                                    opacity=opacity,
+                                    line_width=1,
+                                    line_color='black',
+                                    layer=layer))
+
+    def initialize_shapes(self):
+        """Make graph.layout.shapes for each mne.io.raw.annotation."""
+        if self.inst:
+            ids = [str(uuid1()) for _ in range(len(self.inst.annotations))]
+
+            annots = {id_: {"shape": self._get_annot_shape(annot, name=id_),
+                            "description": self._get_annot_text(annot,
+                                                                name=id_)
+                            }
```

Comment:
As discussed on slack, I think we could have :
```python
class Annotation:

    def __init__(self, onset, duration, description, layout):
        self.id = str(uuid1())
        self.onset = onset
        self.duration = duration
        self._description = description
        self.layout = layout

    def to_mne_annotation(self):
        return mne.Annotations(self.onset, self.duration, self.description)

    # replacing  def _get_annot_shape(self, annotation):
    @property
    def shape(self) : 
        return dict(name=self.id,
                    type="rect",
                    xref="x",
                    yref="y",
                    x0=self.onset,
                    y0=self.layout.yaxis['range'][0],
                    x1=self.onset + self.duration,
                    y1=self.layout.yaxis['range'][1],
                    fillcolor='red',
                    opacity=0.25 if self.duration else .75,
                    line_width=1,
                    line_color='black',
                    layer="below" if self.duration else 'above')

    @shape.setter
    def shape(self, shape_) : 
        self.onset = shape_.x0
        self.duration = shape_.x1 - shape_.x0


    #replacing _get_annot_text(self, annotation):
    @property
    def description(self):
        return dict(x=self.onset + self.duration / 2,
                    y=self.layout.yaxis['range'][1],
                    text=self.description,
                    showarrow=False,
                    yshift=10,
                    font={'color': '#F1F1F1'})
    
    @description.setter
    def description(self, description_str):
        self._description = description_str    

    @staticmethod
    def from_mne_annoration(annot):
        return Annotation(annot["onset"], annot["duration"], annot["description"])

```
Then, here:
```python
annots = [Annotation.from_mne_annotation(annot) for annot in self.inst.annotations]
self.mne_annots.data = {annot.id: annot for annot in annots}
```

and when we want to save back to the raw object, something like:
```python
self.inst.annotations = [annot.to_mne_annotation() for annot in self.mne_annots.data]
```

Note that by moving some of this logic in the Annotation class, we remove this from the MNEVisualizer... which I think is good.

Reply:
I'm +1 for this!

Reply:
> Note that by moving some of this logic in the Annotation class, we remove this from the MNEVisualizer... which I think is good.

Which part of this should be moved out of MNEViz?

Reply:
> I'm +1 for this!

OK. I'll make a commit on that.

> Which part of this should be moved out of MNEViz?

I did not meant out of `mne_visualizer.py` (also it could also live in its own file since it is its own class) but out of the class `MNEVisualizer`. Since this is a class, the code it contains (which is currently partly in `MNEVisualizer` methods) would move out of `MNEVisualizer` since it would be in `Annotation`.  

Just with respect to naming, I think I'll call this class `EEGAnnotation`, just to make it a bit more specific and less likely to be confused with similar classes in MNE or Dash.